ttax/ops.py

```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
+                     'broadcasting is not available.')
+  # batches are not supported yet
+
+  if tt_a.is_tt_matrix:
+    tt_cores = _add_matrix_cores(tt_a, tt_b)
+    return TTMatrix(tt_cores)
+  else:
+    tt_cores = _add_tensor_cores(tt_a, tt_b)
+    return TT(tt_cores)
+
+
+def _add_tensor_cores(tt_a, tt_b):
+  """Internal function to be called from add for two TT-tensors.
+  Does the actual assembling of the TT-cores to add two TT-tensors.
+  """
+  num_dims = tt_a.ndim
+  shape = tt_a.raw_tensor_shape
+  a_ranks = tt_a.tt_ranks
+  b_ranks = tt_b.tt_ranks
+  tt_cores = []
+  for core_idx in range(num_dims):
+    a_core = tt_a.tt_cores[core_idx]
+    b_core = tt_b.tt_cores[core_idx]
+    if core_idx == 0:
+      curr_core = jnp.concatenate((a_core, b_core), axis=2)
+    elif core_idx == num_dims - 1:
+      curr_core = jnp.concatenate((a_core, b_core), axis=0)
+    else:
+      upper_zeros = jnp.zeros((a_ranks[core_idx], shape[core_idx],
+                              b_ranks[core_idx + 1]))
+      lower_zeros = jnp.zeros((b_ranks[core_idx], shape[core_idx],
+                              a_ranks[core_idx + 1]))
+      upper = jnp.concatenate((a_core, upper_zeros), axis=2)
+      lower = jnp.concatenate((lower_zeros, b_core), axis=2)
+      curr_core = jnp.concatenate((upper, lower), axis=0)
+    tt_cores.append(curr_core)
+  return tt_cores
+
+
+def _add_matrix_cores(tt_a, tt_b):
+  """Internal function to be called from add for two TT-matrices.
+  Does the actual assembling of the TT-cores to add two TT-matrices.
+  """
+  num_dims = tt_a.ndim
+  shape = tt_a.raw_tensor_shape
+  a_ranks = tt_a.tt_ranks
+  b_ranks = tt_b.tt_ranks
+  tt_cores = []
+  for core_idx in range(num_dims):
+    a_core = tt_a.tt_cores[core_idx]
+    b_core = tt_b.tt_cores[core_idx]
+    if core_idx == 0:
+      curr_core = jnp.concatenate((a_core, b_core), axis=3)
+    elif core_idx == num_dims - 1:
+      curr_core = jnp.concatenate((a_core, b_core), axis=0)
+    else:
+      upper_zeros = jnp.zeros((a_ranks[core_idx], shape[core_idx],
+                              shape[1][core_idx], b_ranks[core_idx + 1]))
+      lower_zeros = jnp.zeros((b_ranks[core_idx], shape[core_idx],
+                              shape[1][core_idx], a_ranks[core_idx + 1]))
+      upper = jnp.concatenate((a_core, upper_zeros), axis=3)
+      lower = jnp.concatenate((lower_zeros, b_core), axis=3)
+      curr_core = jnp.concatenate((upper, lower), axis=0)
+    tt_cores.append(curr_core)
+  return tt_cores
+
+
+def are_shapes_equal(tt_a, tt_b):
+  """Returns the result of equality check of 2 tensors' shapes: 
+  True if shapes are equal and False otherwise.
+  The arguments should be both TT-tensors or both TT-matrices.
+  The arguments should have the same tensor shape.
+  e.g. for TT core (*, *, ..., *, r_{i}, n_{i}, r_{i+1}):
+  axis *, *, ..., * correspond to the batch shape,
+  axis r_{i}, n_{i}, r_{i+1} correspond to the tensor shape.
```

Comment:
I'm confused by this. Maybe just

The arguments should have the same tensor shape but potentially different TT-ranks

Instead of these four lines?